File: synara/src/app/features/matrix-dto/typing.ts
/**
 * Typing notification DTO.
 */

import type { RoomId, UserId } from './ids';
import { hasForbiddenWireFields, isObject, reqString, stringArray } from './parseUtil';

export type TypingSnapshot = {
  roomId: RoomId;
  userIds: UserId[];
};

export function parseTypingSnapshot(value: unknown): TypingSnapshot | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  const roomId = reqString(value, 'roomId');
  const userIds = stringArray(value.userIds);
  if (roomId === null || userIds === null) return null;
  return { roomId, userIds };
}

File: crates/synara-core/src/dto/typing.rs
//! Typing notification DTO — bounded currently-typing user list.

use serde::{Deserialize, Serialize};

use super::ids::{RoomId, UserId};

/// Currently typing users for a room (product projection).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TypingSnapshot {
    pub room_id: RoomId,
    /// Bounded list of user ids currently typing.
    pub user_ids: Vec<UserId>,
}
